catroid/src/main/java/org/catrobat/catroid/io/DeviceVariableAccessor.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.io;
+
+import android.util.Log;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+
+import org.catrobat.catroid.content.Sprite;
+import org.catrobat.catroid.formulaeditor.UserVariable;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.UUID;
+
+import static org.catrobat.catroid.common.Constants.DEVICE_VARIABLE_JSON_FILENAME;
+
+public final class DeviceVariableAccessor {
+
+	public static final String TAG = DeviceVariableAccessor.class.getSimpleName();
+	private static final int MAX_GENERATE_KEY_TRIES = 100;
+	private static final Object LOCK = new Object();
+	private File deviceVariablesFile;
+
+	public DeviceVariableAccessor(File projectDirectory) {
+		deviceVariablesFile = new File(projectDirectory, DEVICE_VARIABLE_JSON_FILENAME);
+	}
+
+	public boolean readUserVariableValue(UserVariable variable) throws IOException {
+		if (checkDeviceVariableKeyEmpty(variable)) {
+			return false;
+		}
+
+		if (!deviceVariablesFile.exists()) {
+			return false;
+		}
+
+		synchronized (LOCK) {
+			HashMap deviceVariableMap = readMapFromJson();
+
+			if (deviceVariableMap == null) {
+				return false;
+			}
+
+			if (!deviceVariableMap.containsKey(variable.getDeviceValueKey())) {
+				return false;
+			}
+			variable.setValue(deviceVariableMap.get(variable.getDeviceValueKey()));
+		}
+		return true;
+	}
+
+	public void writeVariable(UserVariable userVariable) throws IOException {
+		if (checkDeviceVariableKeyEmpty(userVariable)) {
+			addKeyToUserVariable(userVariable);
+		}
+
+		if (!deviceVariablesFile.exists()) {
+			deviceVariablesFile.createNewFile();
+		}
+
+		synchronized (LOCK) {
+			HashMap deviceVariableMap = readMapFromJson();
+
+			if (deviceVariableMap == null) {
+				deviceVariableMap = new HashMap<>();
+			}
+
+			deviceVariableMap.put(userVariable.getDeviceValueKey(), userVariable.getValue());
+			writeMapToJson(deviceVariableMap);
+		}
+	}
+
+	private UUID generateNewKey() {
+		UUID key = UUID.randomUUID();
+
+		synchronized (LOCK) {
+			HashMap deviceVariableMap;
+
+			try {
+				deviceVariableMap = readMapFromJson();
+			} catch (FileNotFoundException e) {
+				return key;
+			}
+
+			if (deviceVariableMap == null) {
+				return key;
+			}
+			for (int tries = 0; deviceVariableMap.containsKey(key) && tries < MAX_GENERATE_KEY_TRIES; tries++) {
```

Comment:
why the retries?
https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions
"For example, the number of random version-4 UUIDs which need to be generated in order to have a 50% probability of at least one collision is 2.71 quintillion"

Reply:
There is a chance ^^
If you say so, I will remove this function.